File: pkgs/crate-genotype-compiler/src/compiler/remote.rs
use crate::prelude::internal::*;

/// Remote compiler.
pub struct GtcRemote {
    state: GtCompilerState,
    backend: GtbRemote,
}

impl GtcRemote {
    pub async fn handle_runtime_request(
        &mut self,
        request: GtcRemoteRuntimeRequest,
    ) -> Result<GtcRemoteRuntimeRequestResponse> {
        match request {
            GtcRemoteRuntimeRequest::LoadInProject(GtcRemoteRuntimeRequestLoadInProject {
                ..
            }) => {
                let meta = self.load_in_project().await?;
                Ok(GtcRemoteRuntimeRequestResponse::LoadInProject(
                    GtcRemoteRuntimeRequestResponseLoadInProject { meta },
                ))
            }

            GtcRemoteRuntimeRequest::LoadInModules(GtcRemoteRuntimeRequestLoadInModules {
                ..
            }) => {
                let meta = self.load_in_modules().await?;
                Ok(GtcRemoteRuntimeRequestResponse::LoadInModules(
                    GtcRemoteRuntimeRequestResponseLoadInModules { meta },
                ))
            }

            GtcRemoteRuntimeRequest::Compile(GtcRemoteRuntimeRequestCompile { .. }) => {
                let meta = self.compile().await?;
                Ok(GtcRemoteRuntimeRequestResponse::Compile(
                    GtcRemoteRuntimeRequestResponseCompile { meta },
                ))
            }
        }
    }
}

impl GtCompiler<GtcRemoteProps> for GtcRemote {
    fn new(props: GtcRemoteProps) -> Result<Self> {
        let (interop, cwd_path, base_path, config_path) = props;

        let backend = GtbRemote::new(interop, cwd_path, base_path);

        Ok(Self {
            state: GtCompilerState::New {
                config_path,
                meta: GtcMetaNew {},
            },
            backend,
        })
    }

    fn state(&self) -> &GtCompilerState {
        &self.state
    }

    fn state_mut(&mut self) -> &mut GtCompilerState {
        &mut self.state
    }

    fn backend(&self) -> &impl GtBackend {
        &self.backend
    }
}

pub type GtcRemoteProps = (
    Box<dyn GtbRemoteInterop>,
    GtpCwdPath,
    GtpCwdRelativePath,
    Option<GtpCwdRelativePath>,
);

File: pkgs/npm-genotype-types/src/compiler/remote.ts
import { z } from "zod";
import { GtcMetaLoadedProject, GtcMetaLoadedModules, GtcMetaCompiled } from "./meta.js";

export const GtcRemoteRuntimeRequestLoadInProject = z.object({
  kind: z.literal("load-in-project"),
});

export type GtcRemoteRuntimeRequestLoadInProject = z.infer<
  typeof GtcRemoteRuntimeRequestLoadInProject
>;

export const GtcRemoteRuntimeRequestResponseLoadInProject = z.object({
  kind: z.literal("load-in-project"),
  meta: GtcMetaLoadedProject,
});

export type GtcRemoteRuntimeRequestResponseLoadInProject = z.infer<
  typeof GtcRemoteRuntimeRequestResponseLoadInProject
>;

export const GtcRemoteRuntimeRequestLoadInModules = z.object({
  kind: z.literal("load-in-modules"),
});

export type GtcRemoteRuntimeRequestLoadInModules = z.infer<
  typeof GtcRemoteRuntimeRequestLoadInModules
>;

export const GtcRemoteRuntimeRequestResponseLoadInModules = z.object({
  kind: z.literal("load-in-modules"),
  meta: GtcMetaLoadedModules,
});

export type GtcRemoteRuntimeRequestResponseLoadInModules = z.infer<
  typeof GtcRemoteRuntimeRequestResponseLoadInModules
>;

export const GtcRemoteRuntimeRequestCompile = z.object({
  kind: z.literal("compile"),
});

export type GtcRemoteRuntimeRequestCompile = z.infer<typeof GtcRemoteRuntimeRequestCompile>;

export const GtcRemoteRuntimeRequest = z.union([
  GtcRemoteRuntimeRequestLoadInProject,
  GtcRemoteRuntimeRequestLoadInModules,
  GtcRemoteRuntimeRequestCompile,
]);

export type GtcRemoteRuntimeRequest = z.infer<typeof GtcRemoteRuntimeRequest>;

export const GtcRemoteRuntimeRequestResponseCompile = z.object({
  kind: z.literal("compile"),
  meta: GtcMetaCompiled,
});

export type GtcRemoteRuntimeRequestResponseCompile = z.infer<
  typeof GtcRemoteRuntimeRequestResponseCompile
>;

export const GtcRemoteRuntimeRequestResponse = z.union([
  GtcRemoteRuntimeRequestResponseLoadInProject,
  GtcRemoteRuntimeRequestResponseLoadInModules,
  GtcRemoteRuntimeRequestResponseCompile,
]);

export type GtcRemoteRuntimeRequestResponse = z.infer<typeof GtcRemoteRuntimeRequestResponse>;
